CollaboraOnline/collabora-drupal
# Issue #28: docker-compose, github action, and one test.

Fixes #28.

Changes:
- Local Drupal installation with docker-compose, including a Collabora container, so that:
  - Site builders get a local demo to evaluate the module.
  - Contributors can use the local installation to test new functionality.
- A new test case that tests the embedded Collabora editor in readonly mode.
  - This is an ExistingSite* test, using the existing testing/demo installation. It turns out that regular browser tests are difficult to do for this purpose.
- A github action to run all the tests.

### Testing instructions
Follow the updated steps for "local installation" in the README.md.

_If the editor does not open, you may want to cherry-pick the commit from [here](https://github.com/CollaboraOnline/collabora-drupal/commit/371205d13bb7f90fc9344d5e5945c4da5c5c056c) to replace `file_get_contents()` with curl._

To test the "StaleElementReference" problem, you can auto-repeat the test until it occurs.
E.g.
```sh
watch -n1 -e -x docker-compose exec web ./vendor/bin/phpunit -vvv --debug tests/src/ExistingSiteJavascript/CollaboraIntegrationTest.php
```
If you remove the last commit, or change the `retryIfStaleElement` to not catch the exception, then the StaleElementReference will appear.

### Known issues
For now, the actions run in the fork repo only.
https://github.com/donquixote/collabora-drupal/actions
I expect this will change once any PR is merged that contains actions.

For now, we use a docker image from `fpfis`, and we rely on `openeuropa/task-runner` to install the demo/testing site.
This follows the same pattern as e.g. https://www.drupal.org/project/description_list_field, which are published on drupal.org, but where the main development happens in github.
We may rethink this in the future.

For now I need to replace the `file_get_contents` by cherry-picking the commit from [here](https://github.com/CollaboraOnline/collabora-drupal/commit/371205d13bb7f90fc9344d5e5945c4da5c5c056c), to make the local install work properly.
I don't want to include the change in this PR.

I tried a version of the matrix with Drupal 11, and with `--prefer-lowest`, but these are failing.
We should do these another time.
